src/hotspot/share/memory/metaspace.cpp

```diff
@@ -581,86 +582,71 @@ bool Metaspace::class_space_is_initialized() {
-  // when decoding.
-
-  // PPC64: smaller heaps up to 2g will be mapped just below 4g. Then the
-  // attempt to place the compressed class space just after the heap fails on
-  // Linux 4.1.42 and higher because the launcher is loaded at 4g
-  // (ELF_ET_DYN_BASE). In that case we reach here and search the address space
-  // below 32g to get a zerobased CCS. For simplicity we reuse the search
-  // strategy for AARCH64.
-
-  static const struct {
-    address from;
-    address to;
-    size_t increment;
-  } search_ranges[] = {
-    {  (address)(4*G),   (address)(32*G),   4*G, },
-    {  (address)(32*G),  (address)(1024*G), (4 << LogKlassAlignmentInBytes) * G },
-    {  nullptr, nullptr, 0 }
-  };
-
-  // Calculate a list of all possible values for the starting address for the
-  // compressed class space.
-  ResourceMark rm;
-  GrowableArray<address> list(36);
-  for (int i = 0; search_ranges[i].from != nullptr; i ++) {
-    address a = search_ranges[i].from;
-    assert(CompressedKlassPointers::is_valid_base(a), "Sanity");
-    while (a < search_ranges[i].to) {
-      list.append(a);
-      a +=  search_ranges[i].increment;
+// Reserve a range of memory that is to contain narrow Klass IDs. If "strict_base"
+// is true, the memory must be located at an address that can be directly used as
+// encoding base, otherwise the API has more leeway (e.g. optimizing reservation
+// for zero-based encoding).
+ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t size, bool strict_base) {
+
+  char* result = nullptr;
+  const bool randomize = RandomizeClassSpaceLocation;
+
+  // If strict_base, there is no need bothering with low-address reservation since we cannot set
+  // the encoding base to zero.
+  const bool try_in_low_address_ranges = !strict_base;
+
+  // First try to reserve in low address ranges.
+  if (try_in_low_address_ranges) {
+    constexpr uintptr_t unscaled_max = ((uintptr_t)UINT_MAX + 1);
+    log_debug(metaspace, map)("Trying below " SIZE_FORMAT_X " for unscaled narrow Klass encoding", unscaled_max);
+    result = os::attempt_reserve_memory_between(nullptr, (char*)unscaled_max,
+                                                size, Metaspace::reserve_alignment(), randomize);
+    if (result == nullptr) {
+      constexpr uintptr_t zerobased_max = unscaled_max << LogKlassAlignmentInBytes;
+      log_debug(metaspace, map)("Trying below " SIZE_FORMAT_X " for zero-based narrow Klass encoding", zerobased_max);
+      result = os::attempt_reserve_memory_between((char*)unscaled_max, (char*)zerobased_max,
+                                                  size, Metaspace::reserve_alignment(), randomize);
     }
+  } // end: low-address reservation
+
+  if (result == nullptr) {
+    // Failing zero-based allocation, or in strict_base mode, try to come up with
+    // an optimized start address that:
+    // - only has bits set in the third quadrant, to use a single 16-bit move
```

Comment:
It's not clear what "bits set in the third quadrant" and "16-bit move" are.

Also, for removing the old AARCH64/PPC code, perhaps you should do this in a prerequisite PR (taking just this "if" block, plus `os::attempt_reserve_memory_between`). That way we can make sure there's no regression for  AARCH64/PPC, before implementing this full PR.

Reply:
> Also, for removing the old AARCH64/PPC code, perhaps you should do this in a prerequisite PR (taking just this "if" block, plus os::attempt_reserve_memory_between). That way we can make sure there's no regression for AARCH64/PPC, before implementing this full PR.

I would rather avoid that, since it could mean unnecessary churn and an unclearer PR goal.

Replacing the AARCH/PPC code would need both this and the preceding if-block, since PPC piggybacked on the AARCH block for its ability to reserve below 32 GB. 

So that hypothetical PR would need: 
- the full new API, including os::vm_min_address, 
- all gtests, -
-  90% of Metaspace::reserve_address_space_for_compressed_classes

This accounts for the largest part of this patch, there is not much more I could leave out. 

I have tested the patch well, which is currently going through CI at SAP. I am confident in its working as a whole, but I would have to rethink it if we only take parts over. Note that I already left out a lot of changes I originally planned; this is the very much cut-down version.